client/packages/programs/src/RnRForms/DetailView/RnRFormLine.tsx

```diff
@@ -0,0 +1,268 @@
+import React, { useState } from 'react';
+import {
+  BasicTextInput,
+  Checkbox,
+  DatePicker,
+  Formatter,
+  NumericTextInput,
+  useBufferState,
+  useNotification,
+  useTheme,
+  VenCategoryType,
+} from '@openmsupply-client/common';
+import { RnRFormLineFragment } from '../../api/operations.generated';
+
+export const RnRFormLine = ({
+  line,
+  saveLine,
+  periodLength,
+  disabled,
+}: {
+  line: RnRFormLineFragment;
+  periodLength: number;
+  saveLine: (line: RnRFormLineFragment) => Promise<void>;
+  disabled: boolean;
+}) => {
+  const theme = useTheme();
+  const { error } = useNotification();
+
+  const [patch, setPatch] = useState<Partial<RnRFormLineFragment>>({});
+  const draft = { ...line, ...patch };
+
+  const updateDraft = (update: Partial<RnRFormLineFragment>) => {
+    const newPatch = {
+      ...patch,
+      confirmed: false,
+      ...update,
+    };
+
+    const {
+      initialBalance,
+      quantityConsumed,
+      quantityReceived,
+      adjustments,
+      stockOutDuration,
+      previousAverageMonthlyConsumption,
+    } = { ...draft, ...newPatch };
+
+    const finalBalance =
+      initialBalance + quantityReceived - quantityConsumed + adjustments;
+
+    const stockAvailableDays = periodLength - stockOutDuration;
+
+    const adjustedQuantityConsumed = stockAvailableDays
+      ? quantityConsumed * (periodLength / stockAvailableDays)
+      : quantityConsumed;
+
+    // This calculation might be a plugin in future!
+    const averageMonthlyConsumption =
+      // Average of the last 3 months (including the current month)
+      // TODO: what if don't want to consider previous months?
+      (2 * previousAverageMonthlyConsumption + adjustedQuantityConsumed) / 3;
```

Comment:
i.e. on first R&R form, probably DO want AMC to only consider this period...

- make AMC an editable field?
- return list of previous monthly consumption values, so if it's empty we only consider this period? ... Would we store this on the line or query it each time?

Reply:
I like the idea of storing a list of previous average consumption values. then averaging across all available values. I think that's a more accurate result than the (2*previous AMC + current month)/3

unsure how to store the list though. we could have a hard coded count of 3, and just add two 'previous AMC' columns to the table. or a JSON - could we deserialize that to a Vec?

other clients may well want a different number of previous AMCs to be used